tower/src/util/and_then.rs

```diff
@@ -0,0 +1,70 @@
+use std::task::{Context, Poll};
+
+use futures_core::TryFuture;
+use futures_util::{future::AndThen as AndThenFut, TryFutureExt};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Service returned by the [`and_then`] combinator.
+///
+/// [`and_then`]: crate::util::ServiceExt::and_then
+#[derive(Clone, Debug)]
+pub struct AndThen<S, F> {
+    inner: S,
+    f: F,
+}
+
+impl<S, F> AndThen<S, F> {
+    /// Creates a new `AndThen` service.
+    pub fn new(inner: S, f: F) -> Self {
+        AndThen { f, inner }
+    }
+}
+
+impl<S, F, Request, Fut> Service<Request> for AndThen<S, F>
+where
+    S: Service<Request>,
+    F: FnOnce(S::Response) -> Fut + Clone,
+    Fut: TryFuture<Error = S::Error>,
```

Comment:
It's a bit of a bummer that this can't change the error type. I think it might be better to use a `From` bound, and convert the error type returned by `poll_ready`, the way I did in my PR adding a `then` combinator (which is essentially a synchronous version of this), https://github.com/tower-rs/tower/pull/500.

Reply:
@hawkw Adding the `From` bound here is a little more verbose than in `then` because [AndThen](https://docs.rs/futures/0.3.8/futures/future/trait.TryFutureExt.html#method.and_then) constrains the return `Error` type.

We could do this by changing `call` to `self.inner.call(request).and_then(self.f.clone()).map_err(Into::into)` and, as a result, `type Future`. Should I do that?

Reply:
That seems correct to me.

Reply:
Given it a shot.